src/common/metadata.c

```diff
@@ -107,30 +109,31 @@ static void _bulk_remove_metadata(const int img, const gchar *metadata_list)
   }
 }
 
-static void _bulk_add_metadata(const gchar *metadata_list)
+static void _bulk_add_metadata(GList *metadata_list)
 {
   if(metadata_list)
   {
     char *query = NULL;
     sqlite3_stmt *stmt;
-    query = dt_util_dstrcat(query, "INSERT INTO main.meta_data (id, key, value) VALUES %s", metadata_list);
+    query = "INSERT INTO main.meta_data (id, key, value) VALUES (?, ?, ?)";
     DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db), query, -1, &stmt, NULL);
+    sqlite3_bind_text(stmt, 1, g_list_nth_data(metadata_list, 0), -1, 0);
+    sqlite3_bind_text(stmt, 2, g_list_nth_data(metadata_list, 1), -1, 0);
+    sqlite3_bind_text(stmt, 3, g_list_nth_data(metadata_list, 2), -1, 0);
     sqlite3_step(stmt);
     sqlite3_finalize(stmt);
-    g_free(query);
   }
 }
 
 static void _pop_undo_execute(const int imgid, GList *before, GList *after)
 {
   gchar *tobe_removed_list = _get_tb_removed_metadata_string_values(before, after);
-  gchar *tobe_added_list = _get_tb_added_metadata_string_values(imgid, before, after);
+  GList *tobe_added_list = _get_tb_added_metadata_string_values(imgid, before, after);
 
   _bulk_remove_metadata(imgid, tobe_removed_list);
   _bulk_add_metadata(tobe_added_list);
 
   g_free(tobe_removed_list);
-  g_free(tobe_added_list);
```

Comment:
You need to free the list here then:
g_list_free_full(to_be_added_list, g_free)

Reply:
Thanks, I've added them in a new commit